File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/tx/offline/OfflineSigningEnvelopeAdapter.java
package org.hyperledger.iroha.android.tx.offline;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import org.hyperledger.iroha.norito.NoritoAdapters;
import org.hyperledger.iroha.norito.NoritoDecoder;
import org.hyperledger.iroha.norito.NoritoEncoder;
import org.hyperledger.iroha.norito.TypeAdapter;

/** Norito adapter that encodes/decodes {@link OfflineSigningEnvelope} values. */
final class OfflineSigningEnvelopeAdapter implements TypeAdapter<OfflineSigningEnvelope> {

  private static final TypeAdapter<String> STRING_ADAPTER = NoritoAdapters.stringAdapter();
  private static final TypeAdapter<Long> UINT64_ADAPTER = NoritoAdapters.uint(64);
  private static final TypeAdapter<byte[]> BYTES_ADAPTER = NoritoAdapters.bytesAdapter();
  private static final TypeAdapter<Map<String, String>> METADATA_ADAPTER =
      NoritoAdapters.map(STRING_ADAPTER, STRING_ADAPTER);
  private static final TypeAdapter<Optional<byte[]>> OPTIONAL_BYTES_ADAPTER =
      NoritoAdapters.option(BYTES_ADAPTER);

  @Override
  public void encode(final NoritoEncoder encoder, final OfflineSigningEnvelope value) {
    STRING_ADAPTER.encode(encoder, value.schemaName());
    STRING_ADAPTER.encode(encoder, value.keyAlias());
    UINT64_ADAPTER.encode(encoder, value.issuedAtMs());
    METADATA_ADAPTER.encode(encoder, value.metadata());
    BYTES_ADAPTER.encode(encoder, value.encodedPayload());
    BYTES_ADAPTER.encode(encoder, value.signature());
    BYTES_ADAPTER.encode(encoder, value.publicKey());
    OPTIONAL_BYTES_ADAPTER.encode(encoder, value.exportedKeyBundle());
  }

  @Override
  public OfflineSigningEnvelope decode(final NoritoDecoder decoder) {
    final String schemaName = STRING_ADAPTER.decode(decoder);
    final String keyAlias = STRING_ADAPTER.decode(decoder);
    final long issuedAtMs = UINT64_ADAPTER.decode(decoder);
    final Map<String, String> metadata = new LinkedHashMap<>(METADATA_ADAPTER.decode(decoder));
    final byte[] payload = BYTES_ADAPTER.decode(decoder);
    final byte[] signature = BYTES_ADAPTER.decode(decoder);
    final byte[] publicKey = BYTES_ADAPTER.decode(decoder);
    final Optional<byte[]> exportedKey = OPTIONAL_BYTES_ADAPTER.decode(decoder);

    final OfflineSigningEnvelope.Builder builder =
        OfflineSigningEnvelope.builder()
            .setSchemaName(schemaName)
            .setKeyAlias(keyAlias)
            .setIssuedAtMs(issuedAtMs)
            .setMetadata(metadata)
            .setEncodedPayload(payload)
            .setSignature(signature)
            .setPublicKey(publicKey);
    exportedKey.ifPresent(builder::setExportedKeyBundle);
    return builder.build();
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/tx/offline/OfflineSigningEnvelopeAdapter.kt
package org.hyperledger.iroha.sdk.tx.offline

import java.util.LinkedHashMap
import java.util.Optional
import org.hyperledger.iroha.sdk.norito.NoritoAdapters
import org.hyperledger.iroha.sdk.norito.NoritoDecoder
import org.hyperledger.iroha.sdk.norito.NoritoEncoder
import org.hyperledger.iroha.sdk.norito.TypeAdapter

/** Norito adapter that encodes/decodes `OfflineSigningEnvelope` values. */
internal class OfflineSigningEnvelopeAdapter : TypeAdapter<OfflineSigningEnvelope> {

    override fun encode(encoder: NoritoEncoder, value: OfflineSigningEnvelope) {
        STRING_ADAPTER.encode(encoder, value.schemaName)
        STRING_ADAPTER.encode(encoder, value.keyAlias)
        UINT64_ADAPTER.encode(encoder, value.issuedAtMs)
        METADATA_ADAPTER.encode(encoder, value.metadata)
        BYTES_ADAPTER.encode(encoder, value.encodedPayload)
        BYTES_ADAPTER.encode(encoder, value.signature)
        BYTES_ADAPTER.encode(encoder, value.publicKey)
        OPTIONAL_BYTES_ADAPTER.encode(encoder, Optional.ofNullable(value.exportedKeyBundle))
    }

    override fun decode(decoder: NoritoDecoder): OfflineSigningEnvelope {
        val schemaName = STRING_ADAPTER.decode(decoder)
        val keyAlias = STRING_ADAPTER.decode(decoder)
        val issuedAtMs = UINT64_ADAPTER.decode(decoder)
        val metadata = LinkedHashMap(METADATA_ADAPTER.decode(decoder))
        val payload = BYTES_ADAPTER.decode(decoder)
        val signature = BYTES_ADAPTER.decode(decoder)
        val publicKey = BYTES_ADAPTER.decode(decoder)
        val exportedKey: Optional<ByteArray> = OPTIONAL_BYTES_ADAPTER.decode(decoder)

        return OfflineSigningEnvelope(
            encodedPayload = payload,
            signature = signature,
            publicKey = publicKey,
            schemaName = schemaName,
            keyAlias = keyAlias,
            issuedAtMs = issuedAtMs,
            metadata = metadata,
            exportedKeyBundle = exportedKey.orElse(null),
        )
    }

    companion object {
        private val STRING_ADAPTER: TypeAdapter<String> = NoritoAdapters.stringAdapter()
        private val UINT64_ADAPTER: TypeAdapter<Long> = NoritoAdapters.uint(64)
        private val BYTES_ADAPTER: TypeAdapter<ByteArray> = NoritoAdapters.bytesAdapter()
        private val METADATA_ADAPTER: TypeAdapter<Map<String, String>> =
            NoritoAdapters.map(STRING_ADAPTER, STRING_ADAPTER)
        private val OPTIONAL_BYTES_ADAPTER: TypeAdapter<Optional<ByteArray>> =
            NoritoAdapters.option(BYTES_ADAPTER)
    }
}
